clients/python-wrapper/lakefs/object.py

```diff
@@ -2,19 +2,378 @@
+            self._client.sdk_client.objects_api.copy_object(repository=self._repo_id,
+                                                            branch=destination_branch_id,
+                                                            dest_path=destination_path,
+                                                            object_copy_creation=object_copy_creation)
+
+        return WriteableObject(repository=self._repo_id, reference=destination_branch_id, path=destination_path,
+                               client=self._client)
+
+
+class ObjectReader:
+    """
+    ReadableObject provides read-only functionality for lakeFS objects with IO semantics.
+    This Object is instantiated and returned on open() methods for immutable reference types (Commit, Tag...)
+    """
+    _client: Client
+    _obj: StoredObject
+    _mode: OpenModes
+    _pos: int
+    _pre_sign: Optional[bool] = None
+
+    def __init__(self, obj: StoredObject, mode: OpenModes, pre_sign: Optional[bool] = None,
+                 client: Optional[Client] = DefaultClient) -> None:
+        if mode not in get_args(OpenModes):
+            raise ValueError(f"invalid read mode: '{mode}'. ReadModes: {OpenModes}")
+
+        self._obj = obj
+        self._mode = mode
+        self._pre_sign = pre_sign if pre_sign is not None else client.storage_config.pre_sign_support
+        self._client = client
+        self._pos = 0
+
+    def tell(self) -> int:
+        """
+        Object's read position. If position is past object byte size, trying to read the object will return EOF
+        """
+        return self._pos
+
+    def seek(self, pos) -> None:
+        """
+        Move the object's reading position
+        :param pos: The position (in bytes) to move to
+        :raises OSError if provided position is negative
+        """
+        if pos < 0:
+            raise OSError("position must be a non-negative integer")
+        self._pos = pos
+
+    def read(self, read_bytes: int = None) -> TextIO | BinaryIO:
+        """
+        Read object data
+        :param read_bytes: How many bytes to read. If read_bytes is None, will read from current position to end.
+        If current position + read_bytes > object size.
+        :return: The bytes read
+        :raises
+            EOFError if current position is after object size
+            OSError if read_bytes is non-positive
+            ObjectNotFoundException if repo id, reference id or object path does not exist
+            PermissionException if user is not authorized to perform this operation, or operation is forbidden
+            ServerException for any other errors
+        """
+        if read_bytes and read_bytes <= 0:
+            raise OSError("read_bytes must be a positive integer")
+
+        end_pos = self._pos + read_bytes - 1 if read_bytes is not None else ""
+        read_range = _RANGE_STR_TMPL.format(start=self._pos, end=end_pos)
```

Comment:
1. end_pos is sometimes a number and sometimes a string - suggest to use None.
2. we always format a read range for the get request. we can make it optional if we also check the start case.

```python
def get_range_string(start, end=None):
    if start == 0 and end is not None:
        return None
    if end is None:
        return f"bytes={start}-"
    return f"bytes={start}-{end}"
```

Reply:
1. None formats to the literal 'None'
2. 👍🏽